spring-cloud-commons/src/main/java/reactor/core/publisher/CloudFlux.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2019-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.core.publisher;
+
+import org.reactivestreams.Publisher;
+
+/**
+ * INTERNAL USAGE ONLY. This functionality is ported to reactor-core and will be removed
+ * in a next release.
+ *
+ * @author Tim Ysewyn
+ */
+public abstract class CloudFlux<T> extends Flux<T> {
+
+	/**
+	 * Pick the first {@link Publisher} to emit an onNext/onError signal and replay all
+	 * signals from that {@link Publisher}, effectively behaving like the fastest of these
+	 * competing sources. If none of the sources is emitting a signal, a completion signal
```

Comment:
I'd rephrase a bit: "If all the source complete empty, a single completion signal is sent. Note that if all the sources are empty (never emit an element, ie. no onNext) AND at least one is also infinite (no onComplete/onError signal), the resulting `Flux` will be infinite and empty (like `Flux#never()`)."

Reply:
to formalize the 3 cases I'm covering above:
```java
CloudFlux.firstNonEmpty(Flux.never(), Flux.empty(), Flux.just("foo")); //emits foo & completes
CloudFlux.firstNonEmpty(Flux.empty(), Flux.empty()); //completes
CloudFlux.firstNonEmpty(Flux.never(), Flux.empty()); //infinite nothing
```